base/ltcmd.dtx

```diff
@@ -3002,68 +3067,193 @@
+    \bool_set_false:N \l_@@_obey_spaces_bool
+    \@@_grab_f_aux:n {#1}
   }
-\cs_new_protected:Npn \@@_grab_D_long:w #1#2#3 \@@_run_code:
+\cs_new_protected:Npn \@@_grab_f_obey_spaces:w #1 \@@_run_code:
   {
-    \@@_grab_D_aux:NNnNNN #1 #2 {#3} \cs_set_protected:Npn
-      \@@_peek_nonspace_remove:NTF \use_ii:nn
+    \bool_set_true:N \l_@@_obey_spaces_bool
+    \@@_grab_f_aux:n {#1}
   }
-\cs_new_protected:Npn \@@_grab_D_obey_spaces:w #1#2#3 \@@_run_code:
+\cs_new_protected:Npn \@@_grab_f_aux:n #1
   {
-    \@@_grab_D_aux:NNnNNN #1 #2 {#3} \cs_set_protected_nopar:Npn
-      \@@_peek_meaning_remove:NTF \use_ii:nn
+    \tl_set:Nn \l_@@_signature_tl {#1}
+    \group_begin:
+      \tl_clear:N \l_@@_v_arg_tl
+      \tex_escapechar:D = 92 \scan_stop:
+      \tex_endlinechar:D = `\^^M \scan_stop:
+      \cs_set_eq:NN \do \char_set_catcode_other:N
+      \dospecials
+      \tl_set:Ne \l_@@_tmpa_tl
+        {
+          \c_backslash_str end
+            \c_left_brace_str \@currenvir \c_right_brace_str
+        }
+      \use:e
+        {
+          \cs_set_protected:Npn \@@_grab_f_check:w
+            ##1 \l_@@_tmpa_tl ##2 \l_@@_tmpa_tl ##3
+            \scan_stop:
+        }
+        {
+          \tl_if_empty:nTF {##3}
+            {
+              \tl_put_right:Nn \l_@@_v_arg_tl
+                {
+                  ##1
+                  \obeyedline
+                }
+              \@@_grab_f_loop:w
+            }
+            { \@@_grab_f_end:nn {##1} {##2} }
+        }
+      \@@_grab_f_aux:w
   }
-\cs_new_protected:Npn \@@_grab_D_long_obey_spaces:w #1#2#3 \@@_run_code:
-  {
-    \@@_grab_D_aux:NNnNNN #1 #2 {#3} \cs_set_protected:Npn
-      \@@_peek_meaning_remove:NTF \use_ii:nn
+%    \end{macrocode}
+%   To allow for the need to change the category code of end-of-lines, we
+%   put this in a small auxiliary.
+%    \begin{macrocode}
+\group_begin:
+  \char_set_catcode_other:N \^^M %
+  \cs_new_protected:Npn \@@_grab_f_aux:w %
+    { %
+      \cs_set_protected:Npe \@@_grab_f_loop:w ##1 ^^M %
+        { %
+          \@@_grab_f_check:w ##1 %
+            \l_@@_tmpa_tl %
+            \l_@@_tmpa_tl %
+            \scan_stop: %
+        } %
+      \char_set_catcode_other:N \^^M %
+      \@@_grab_f_loop:w %
+    } %
+\group_end: %
+\cs_new_protected:Npn \@@_grab_f_check:w { }
+\cs_new_protected:Npn \@@_grab_f_loop:w { }
+\cs_new_protected:Npn \@@_grab_f_end:nn #1#2
+  {
+      \tl_if_blank:nF {#1#2}
+        \ERROR
+    \exp_args:NNNo \group_end:
+    \tl_set:Nn \l_@@_v_arg_tl { \l_@@_v_arg_tl }
+    \@@_add_arg:x
+      {
+        \bool_if:NTF \l_@@_obey_spaces_bool
+          { \exp_not:V }
+          { \exp_args:NV \@@_grab_f_end:n }
+          \l_@@_v_arg_tl
+      }
+    \exp_args:NV \end \@currenvir
   }
-\cs_new_protected:Npn \@@_grab_D_no_strip:w
-  #1#2#3 \@@_run_code:
+%    \end{macrocode}
+%   Look for line markers at each end and tidy up if required.
+%    \begin{macrocode}
+\cs_new:Npn \@@_grab_f_end:n #1
```

Comment:
I'm not sure what you are getting at?

Reply:
Your current `\obeyedline`-stripping code tests for a leading `\obeyedline` then removes it, tests for a trailing `\obeyedline` then removes it. This here just removes the leading `\obeyedline` (and if there is none it removes the one after the second `\q_nil`, then removes the trailing one (and if there is none it removes the one in front of the third `\q_nil`, and then removes the markers and the remainder of the fallback. Reads much clearer and is faster than your current code to strip them. This suggestion should replace all of your `\@@_grab_f_end:n` code not only that single line (I could not mark the suggestion to cover the whole range because there were deleted lines interspersed in the diff and GH doesn't like that).

Reply:
Changed